src/commandline_frame.ts

```diff
@@ -256,14 +257,14 @@ export function clear(evlistener = false) {
 commandline_state.clear = clear
 
 /** @hidden **/
-function history(n) {
+async function history(n) {
```

Comment:
This introduces race conditions. Without your commit, every function in `lib/comandline_cmds.ts` is sync, this means that any command that changes the commandline input's value (e.g. `prev_history`) is guaranteed to complete before a new command is issued. Making `history` async (and `complete`, in `lib/commandline_cmds.ts`) means that calling these twice really fast could result in surprising results, e.g. `<ArrowUp><ArrowUp>` could show command N-1 rather than command N-2.

Reply:
Yeah, I noticed this. The right way to solve this is to have each one cancel all the others, and to keep a local counter for history position in a not async portion.

However, in practice, I don't think this will be a massive problem as it should just mean that holding up navigates command history more slowly (presumably to the point where on a slow machine it stops navigating at all). What do you think?

Reply:
>The right way to solve this is to have each one cancel all the others, and to keep a local counter for history position in a not async portion.

In my opinion we should aim for a mechanism a bit more general than that, otherwise we're going to encounter issues when we'll want to implement async commands that have nothing to do with history.

Reply:
In practice this seems fine for now. Tested in the background console with:

`tri.state.cmdHistory = Array.from({length: 10000},(v, i) => i + "")`

It slows down if the history is around 100,000+ items but I think that's probably a problem for another day, given that my main profile only has 2500 items in it currently. It never appears out of order.